### PR TITLE
Add ERNIE 3.0 performance test

### PR DESCRIPTION
**Pull request by LiuChiachi**
### PR types

New  features
### PR changes

Models
### Description

Add ERNIE 3.0 performance test

后端更新PR：https://github.com/PaddlePaddle/PaddleNLP/pull/2780
GPU：
FP32：ort
FP16：trt换为ort【这个可能和表格里性能数据不match】
Int8：trt
CPU：
FP32：ort换mkldnn，但是差不多
Int8：ORT动态量化，现在使用MKLDNN支持离线量化

折线图GPU都用的是paddle inference
CPU用的是ONNXRuntime

### CODE REVIEW

File: model_zoo/ernie-3.0/performance_test/README.md [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+性能测试需要静态图模型 ${MODEL_PATH}，包含 `*.pdmodel` 和 `*.pdiparams` 文件。动态图转静态图可参考[预训练模型导出脚本](../../bert/export_model.py)
+
+这里以 CLUE 中的 IFLYTEK 数据集为例，因为 IFLYTEK 是长文本数据集，平均长度是 289.17（测试脚本默认的最大序列长度是 128）。
+
+## GPU
+
+### 环境依赖
+
+GPU 下的推理可以使用 TensorRT 加速，可以安装[包含 TensorRT 预测库的 PaddlePaddle](https://www.paddlepaddle.org.cn/inference/user_guides/download_lib.html)。需要注意的是，量化模型的预测必须安装包含 TensorRT 预测库的 PaddlePaddle。
```
Comment (tianxin1860) [inline model_zoo/ernie-3.0/performance_test/README.md]: 咱们 README 中的 GPU 性能是基于 TensorRT 加速来测试, 而不是基于 onnxruntime-gpu 来测试么？

File: model_zoo/ernie-3.0/performance_test/README.md [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+### 启动测试
+
+推理精度默认 PF32，针对 FP16 或者 INT 的推理需要传入对应参数 `--fp16` 或者 `--int8`。还可以传入 `batch_size` 等参数。如：
+
+```shell
+MODEL_NAME_OR_PATH=ernie-3.0-medium-zh
+bs=32
+# Step1: collect shape info
```
Comment (tianxin1860) [inline model_zoo/ernie-3.0/performance_test/README.md]: collect shape info 是指针对量化模型需要吧？  还是说 fp32、fp16、int8 都需要？

File: model_zoo/ernie-3.0/performance_test/README.md [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+python clue_infer.py \
+    --model_path ${MODEL_PATH} \
```
Comment (tianxin1860) [inline model_zoo/ernie-3.0/performance_test/README.md]: --model_path 和 --model_name_or_path 这 2 个参数必须同时存在么？用户直观看到这2个参数命名就会比较困惑。

File: model_zoo/ernie-3.0/performance_test/README.md [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    --task_name iflytek --perf \
+    --device gpu  --use_inference \
+    --use_trt --collect_shape \
```
Comment (tianxin1860) [inline model_zoo/ernie-3.0/performance_test/README.md]: 风格保持一致吧，每个参数占一行。

File: model_zoo/ernie-3.0/performance_test/README.md [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+    --faster_tokenizer \
+    --model_name_or_path ${MODEL_NAME_OR_PATH}
+
+# Step2(Option2):precision FP16
+python clue_infer.py \
+    --model_path ${MODEL_PATH} \
+    --task_name iflytek \
+    --perf --device gpu  \
+    --use_inference --use_trt --fp16  \
+    --batch_size $bs \
```
Comment (tianxin1860) [inline model_zoo/ernie-3.0/performance_test/README.md]: 同上。

**Review by tianxin1860 — COMMENTED** [review]
Leave some comments

File: model_zoo/ernie-3.0/performance_test/clue_infer.py [hunk 6]
```diff
@@ -0,0 +1,367 @@
...
+    )
+    parser.add_argument(
+        "--fp16",
+        action='store_true',
+        help="Whether to use int16 inference.",
+    )
+    parser.add_argument(
+        "--use_inference",
+        action='store_true',
+        help="use_inference for inference.",
```
Comment (tianxin1860) [inline model_zoo/ernie-3.0/performance_test/clue_infer.py]: 这个参数命名含义是 使用 paddle inference 做预测还是？

File: model_zoo/ernie-3.0/performance_test/clue_infer.py [hunk 7]
```diff
@@ -0,0 +1,367 @@
...
+@paddle.no_grad()
+def evaluate(outputs, metric, data_loader):
+    metric.reset()
+    for i, batch in enumerate(data_loader):
+        input_ids, segment_ids, labels = batch
+        logits = paddle.to_tensor(outputs[i][0])
+        correct = metric.compute(logits, labels)
+        metric.update(correct)
+    res = metric.accumulate()
+    print("acc: %s, " % res, end='')
```
Comment (tianxin1860) [inline model_zoo/ernie-3.0/performance_test/clue_infer.py]: 为什么需要 evaluate 函数？ 看着像冗余代码。

File: model_zoo/ernie-3.0/performance_test/clue_infer.py [hunk 8]
```diff
@@ -0,0 +1,367 @@
...
+    parser.add_argument(
+        "--collect_shape",
+        action='store_true',
+        help="Whether to collect shape info.",
+    )
+    args = parser.parse_args()
+    return args
+
+
+def convert_example(example,
```
Comment (tianxin1860) [inline model_zoo/ernie-3.0/performance_test/clue_infer.py]: 以 IFLYTEK 作为基准数据集来测性能的话，这个函数数据预处理逻辑应该是简单明了的，不至于这么复杂。

**Review by tianxin1860 — COMMENTED** [review]
数据预处理部分看起来有冗余逻辑，以及 evaluate 这类冗余代码都需要在提 PR 之前意识到这些问题。

File: model_zoo/ernie-3.0/performance_test/README.md [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+本目录提供了 GPU 和 CPU 下性能测试的脚本。
+
+脚本默认不使用 FasterTokenizer 对推理加速，可通过传入参数`--faster_tokenizer`来使用 FasterTokenizer 加速推理，当使用 FasterTokenizer 时，需要先安装 faster_tokenizer 包，并通过设置环境变量 `OMP_NUM_THREADS` 来设置切词的线程数，例如：
+
+```shell
+pip install faster_tokenizer
+export OMP_NUM_THREADS=4
+```
+
+性能测试需要静态图模型 ${MODEL_PATH}，包含 `*.pdmodel` 和 `*.pdiparams` 文件。动态图转静态图可参考[预训练模型导出脚本](../../bert/export_model.py)
```
Comment (tianxin1860) [inline model_zoo/ernie-3.0/performance_test/README.md]: ERNIE 系列模型动转静属于性能测试必选项，建议在本目录提供动转静的脚本，而不是跳出去参考 BERT 动转静。

**Review by tianxin1860 — COMMENTED** [review]
需要增加 ERNIE 系列模型动转静脚本及使用示例。

File: model_zoo/ernie-3.0/performance_test/README.md (hunk 1 above)
Comment (LiuChiachi) [reply model_zoo/ernie-3.0/performance_test/README.md]: 是基于TensorRT 加速。之前的ERNIE3.0发布的GPU测试也都是基于TensorRT 的

File: model_zoo/ernie-3.0/performance_test/README.md (hunk 2 above)
Comment (LiuChiachi) [reply model_zoo/ernie-3.0/performance_test/README.md]: 使用TRT都需要

File: model_zoo/ernie-3.0/performance_test/README.md (hunk 3 above)
Comment (LiuChiachi) [reply model_zoo/ernie-3.0/performance_test/README.md]: 需要，因为前者是静态图模型，不知道模型的`model_name_or_path`，测试脚本需要传入`model_name_or_path`来初始化tokenizer

File: model_zoo/ernie-3.0/performance_test/README.md (hunk 9 above)
Comment (LiuChiachi) [reply model_zoo/ernie-3.0/performance_test/README.md]: 因为那份代码是通用的，ERNIE 的模型用起来也不需要改，所以就直接链接过去了，如果复制一份过来的话好像也挺冗余的，或者要不把这种通用的工具放在一个模型无关的目录里保存着？

**Review by ZeyuChen — COMMENTED** [review]
整体目录名简化为perf更简洁

File: model_zoo/ernie-3.0/performance_test/README.md (hunk 9 above)
Comment (LiuChiachi) [reply model_zoo/ernie-3.0/performance_test/README.md]: 目前用软链来解决这个问题，没有增加重复的代码，用户也更直观。

File: model_zoo/ernie-3.0/performance_test/clue_infer.py (hunk 8 above)
Comment (LiuChiachi) [reply model_zoo/ernie-3.0/performance_test/clue_infer.py]: 感谢指出，这块比较复杂，目前来看没有必要保留，已简化。

File: model_zoo/ernie-3.0/performance_test/clue_infer.py (hunk 7 above)
Comment (LiuChiachi) [reply model_zoo/ernie-3.0/performance_test/clue_infer.py]: 感谢指出，应该是之前不是纯perf代码的遗留，已删除

File: model_zoo/ernie-3.0/performance_test/clue_infer.py (hunk 6 above)
Comment (LiuChiachi) [reply model_zoo/ernie-3.0/performance_test/clue_infer.py]: 是的，这里的描述已更新，避免引起不确定和疑惑

File: model_zoo/ernie-3.0/performance_test/README.md (hunk 5 above)
Comment (LiuChiachi) [reply model_zoo/ernie-3.0/performance_test/README.md]: 感谢指出，已修改。

File: model_zoo/ernie-3.0/performance_test/README.md (hunk 4 above)
Comment (LiuChiachi) [reply model_zoo/ernie-3.0/performance_test/README.md]: 感谢指出，已修改。

**Comment by LiuChiachi** [comment]
> 整体目录名简化为perf更简洁

Thanks. Done.

File: model_zoo/ernie-3.0/perf/clue_infer.py [hunk 10]
```diff
@@ -0,0 +1,270 @@
...
+                                              device=args.device,
+                                              use_int8=args.int8,
+                                              use_fp16=args.fp16,
+                                              collect_shape=args.collect_shape,
+                                              num_threads=args.num_threads,
+                                              use_inference=args.use_inference,
+                                              use_trt=args.use_trt)
+        if args.collect_shape:
+            min_batch_size, max_batch_size, opt_batch_size = 1, args.batch_size, args.batch_size
+            min_seq_len, max_seq_len, opt_seq_len = 2, 128, 32
```
Comment (ZeyuChen) [inline model_zoo/ernie-3.0/perf/clue_infer.py]: 这里的128是否需要跟args.max_seq_len关联好呢？

File: model_zoo/ernie-3.0/perf/clue_infer.py [hunk 11]
```diff
@@ -0,0 +1,270 @@
...
+    )
+    parser.add_argument(
+        "--use_inference",
+        action='store_true',
+        help="use paddle.inference for inference.",
+    )
+    parser.add_argument(
+        "--collect_shape",
+        action='store_true',
+        help="Whether to collect shape info.",
```
Comment (ZeyuChen) [inline model_zoo/ernie-3.0/perf/clue_infer.py]: Whether to collect shape info for TensorRT acceleration.
这个collectshape是否是推荐用法呢？还是我们模型设置的时候可以手动配置出来最佳的选项？

File: model_zoo/ernie-3.0/perf/infer_backend.py [hunk 12]
```diff
@@ -0,0 +1,242 @@
...
+            elif device == "cpu":
+                config.disable_gpu()
+                config.switch_ir_optim(True)
+                config.enable_mkldnn()
+                config.set_cpu_math_library_num_threads(num_threads)
+                paddle.set_device("cpu")
+
+            if use_trt:
+                assert device == "gpu", "when use_trt, the device must be gpu."
+                if use_int8:
```
Comment (ZeyuChen) [inline model_zoo/ernie-3.0/perf/infer_backend.py]: 想这类带嘛，如果这么长一段，只是个参数不一样
是否可以一个改造位
precision_mode=inference.PrecisionType.Float32
if use_int8:
  precition_mode = inference.PrecisionType.Int8
elif use_fp16
 precision_mode = inference.PrecisionType.Half
呢？
然后再将precision_mode变量单独传入enable_ensorrt_engine，这样代码是否就会更加简洁点？

File: model_zoo/ernie-3.0/perf/infer_backend.py [hunk 13]
```diff
@@ -0,0 +1,242 @@
...
+            from paddle import inference
+            import paddle
+            config = paddle.inference.Config(model_path + ".pdmodel",
+                                             model_path + ".pdiparams")
+            config.enable_use_gpu(100, 0)
+            paddle.set_device("gpu")
+            if use_int8:
+                config.enable_tensorrt_engine(
+                    workspace_size=1 << 30,
+                    precision_mode=inference.PrecisionType.Int8,
```
Comment (ZeyuChen) [inline model_zoo/ernie-3.0/perf/infer_backend.py]: 这里也是一样的，大量的冗余配置开关
以及trt的逻辑和trt的逻辑就有点重复和层次不一致
整体树立下代码逻辑关系，应该控制在80行左右完成相关的配置。

File: model_zoo/ernie-3.0/perf/infer_backend.py [hunk 14]
```diff
@@ -0,0 +1,242 @@
...
+                                                  providers=providers)
+            input_name1 = self.predictor.get_inputs()[0].name
+            input_name2 = self.predictor.get_inputs()[1].name
+            self.input_handles = [input_name2, input_name1]
+            self.output_handles = []
+        print(">>> [InferBackend] engine created ...")
+
+    def infer(self, data):
+        if isinstance(self.predictor,
+                      paddle.fluid.core_avx.PaddleInferPredictor):
```
Comment (ZeyuChen) [inline model_zoo/ernie-3.0/perf/infer_backend.py]: 也不应该再出现fluid层级的代码

**Review by ZeyuChen — COMMENTED** [review]
代码整体逻辑做下梳理吧，非常冗余重复，质量较低。

**Comment by LiuChiachi** [comment]
后端更新PR：https://github.com/PaddlePaddle/PaddleNLP/pull/2780
GPU：
FP32：ort
FP16：trt换为ort【这个可能和表格里性能数据不match】
Int8：trt
CPU：
FP32：ort换mkldnn，但是差不多
Int8：ORT动态量化，现在使用MKLDNN支持离线量化

折线图GPU都用的是paddle inference
CPU用的是ONNXRuntime
